pkg/riff/commands/completion.go

```diff
@@ -17,20 +17,54 @@
 package commands
 
 import (
+	"context"
+
 	"github.com/projectriff/riff/pkg/cli"
 	"github.com/spf13/cobra"
 )
 
+type CompletionOptions struct {
+	Shell string
+}
+
+func (opts *CompletionOptions) Validate(ctx context.Context) *cli.FieldError {
+	errs := &cli.FieldError{}
+
+	if opts.Shell == "" {
+		errs = errs.Also(cli.ErrMissingField(cli.ShellFlagname))
+	} else if opts.Shell != "bash" && opts.Shell != "zsh" {
+		errs = errs.Also(cli.ErrInvalidValue(opts.Shell, cli.ShellFlagname))
+	}
+
+	return errs
+}
+
+func (opts *CompletionOptions) Exec(ctx context.Context, c *cli.Config) error {
+	cmd := cli.CommandFromContext(ctx)
+	switch opts.Shell {
+	case "bash":
+		return cmd.Root().GenBashCompletion(c.Stdout)
+	case "zsh":
+		return cmd.Root().GenZshCompletion(c.Stdout)
+	default:
+		panic("invalid shell: " + opts.Shell) // protected by opts.Validate()
+	}
+	return nil
+}
+
 func NewCompletionCommand(c *cli.Config) *cobra.Command {
+	opts := &CompletionOptions{}
+
 	cmd := &cobra.Command{
 		Use:     "completion",
 		Short:   "<todo>",
 		Example: "<todo>",
 		Args:    cli.Args(),
-		Run: func(cmd *cobra.Command, args []string) {
-			cmd.Root().GenBashCompletion(cmd.OutOrStdout())
-		},
+		PreRunE: cli.ValidateOptions(opts),
+		RunE:    cli.ExecOptions(c, opts),
 	}
 
+	cmd.Flags().StringVar(&opts.Shell, cli.StripDash(cli.ShellFlagname), "bash", "shell to generate completion for: bash or zsh")
```

Comment:
If you don't like a flag here, subcommands for bash and zsh would be preferable to an opaque argument. Positional arguments don't make sense to me when the value is from a fixed set, especially n=2.

Reply:
I think I was drawn to arg as a matter of habit. Your argument does make sense, going with a flag